src/query/ts/m3db/consolidator.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3db
+
+import (
+	"math"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/ts"
+	xts "github.com/m3db/m3/src/query/ts"
+)
+
+const initLength = 10
+
+func removeStale(
+	earliestLookback time.Time,
+	dps []ts.Datapoint,
+) []ts.Datapoint {
+	var (
+		i  int
+		dp ts.Datapoint
+	)
+	for _, dp = range dps {
+		// Timestamp is within lookback period
+		if !dp.Timestamp.Before(earliestLookback) {
+			break
+		}
+
+		i++
+	}
+
+	return dps[i:]
+}
+
+type lookbackConsolidator struct {
+	lookback         time.Duration
```

Comment:
what's lookback here ?

Reply:
This is the same lookback as `models.LookbackDelta` for prom consolidation

Reply:
Isn't that the earliestLookback ? Maybe change the names a bit to make it clear ?

Reply:
lookback is the actual duration, `earliestLookback` is the cutoff time for incoming values

Reply:
Renamed to `lookbackDuration`